students/NatalieRodriguez/Lesson06/mailroom_part4.py

```diff
@@ -0,0 +1,160 @@
+
+        options[response][1]()  # Call helper function
+
+
+def exit_dash():
+
+    print("\nExiting the donor dashboard. Goodbye! \n")
+    return
+
+
+def thank_you():
+    """
+    Add new donations for new or existing donors, and send a thank-you
+    letter.
+    :return:  None.
+    """
+    # Get the donor name, show all donors, or quit
+    response = input(
+        "\nType the full donor name (Enter 'list' to show all donors or enter 'exit'): "
+    ).strip()
+
+    if response.lower() in ('', 'exit'):
+        exit_dash()
+        return
+
+    elif response.lower() == 'list':
+        print("\nDONORS:")
+        for donor in donors:
+            print(donor)
+        thank_you()
+
+    else:
+        while True:
+            donation = input(
+                f"Enter amount of donation by '{response}' (or enter 'exit'): "
+            ).strip().lower()
+            if donation == 'exit':
+                exit_dash()
+                return
+
+            elif donation.strip('0123456789.') == '' and len(
+                    donation) > 0 and donation.count('.') <= 1 and float(
+                donation) > 0.0:
+                break
+
+        donation = float(donation)
+        donors.setdefault(response, [])
+        donors[response].append(donation)
+        print(create_form_letter(response, donation))
+
+
+def create_report():
+
+    print('\n')
+    print('        Donor Name        |    Total Given      | '
+          + 'Number of Gifts |    Average Gift    ')
+    print('--------------------------|---------------------|-'
+          + '----------------|--------------------')
+    for individual_donor, donations in donors.items():
+        total_donation = sum(donations)
+        number_of_gifts = len(donations)
+        average_donation = 1.0 * total_donation / number_of_gifts
+        print('{:<25s} | ${:>18,.2f} | {:>15d} | ${:>18,.2f}'.format(
+            individual_donor, total_donation,
+            number_of_gifts, average_donation))
+
+
+def thank_all():
+
+    cur_dir = os.getcwd()
+    print('\nThe current directory is %s' % cur_dir)
+    new_dir = input('\nEnter the directory you want to save to '
+                    '(Press enter to save in the current location): ').strip()
+#exception handling changes here: indented all for try/except block
+    try:
+        if new_dir != '':
+            os.mkdir(new_dir)
+            os.chdir(new_dir)
+            new_dir = os.getcwd()
+        else:
+            new_dir = cur_dir
+
+        for k, v in donors.items():
+            letter = create_form_letter(k, v[-1])
+            with open('{:s}.txt'.format(k), 'w') as f:
+                for line in letter:
+                    f.write(line)
+
+        print('Your thank you notes have been saved in %s:' % new_dir)
+        print(os.listdir())
+        os.chdir(cur_dir)
+    except FileNotFoundError:
+        print('\n'"We cannot find that file location. Please enter another.")
+    finally:
+        thank_all()
```

Comment:
separate file writing into separate function and add a test that calls your function and asserts that each file was created.